File: algorithms/cpp/largestSubstringBetweenTwoEqualCharacters/largestSubstringBetweenTwoEqualCharacters.cpp
// Source: https://leetcode.com/problems/largest-substring-between-two-equal-characters/
// Author: Miao Zhang
// Date:   2021-05-24

class Solution {
public:
    int maxLengthBetweenEqualCharacters(string s) {
        vector<int> m(26, -1);
        int res = -1;
        for (int i = 0; i < s.size(); i++) {
            int idx = s[i] - 'a';
            if (m[idx] == -1) {
                m[idx] = i;
            } else {
                res = max(res, i - m[idx] - 1);
            }
        }
        return res;
    }
};

File: algorithms/python/largestSubstringBetweenTwoEqualCharacters/largestSubstringBetweenTwoEqualCharacters.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/largest-substring-between-two-equal-characters/
# Author: Miao Zhang
# Date:   2021-05-24

class Solution:
    def maxLengthBetweenEqualCharacters(self, s: str) -> int:
        m = {}
        res = -1
        for i, c in enumerate(s):
            if c not in m:
                m[c] = i
            else:
                res = max(res, i - m[c] - 1)
        return res
